File: packages/perception/src/ffmpeg/audio.ts
import { compareText } from '@editorial-ir/contracts';
import type { AnalyzeAudioParams, AnalyzeAudioResult } from '@editorial-ir/contracts';
import type { AudioModel, ModelIdentity } from '../types.js';
import { computeHopStatistics, type HopStatistics } from '../wav.js';

/**
 * Loudness, silence and speech presence, computed from the prepared WAV.
 *
 * This deliberately stops short of classifying laughter, applause or music:
 * those need a trained audio tagger, which belongs in the Python worker. What it
 * does cover is the part the editor actually needs on every machine — where the
 * quiet is — because that is what lets a cut land between words instead of
 * through one.
 */
export interface WavAudioAnalyzerOptions {
  /** A hop below the floor plus this margin counts as silent. */
  silenceMarginDb?: number;
  /** Shortest run that becomes a silence event. */
  minSilenceMs?: number;
  /** Shortest run that becomes a speech event. */
  minSpeechMs?: number;
}

export class WavAudioAnalyzer implements AudioModel {
  readonly identity: ModelIdentity;
  private readonly silenceMarginDb: number;
  private readonly minSilenceMs: number;
  private readonly minSpeechMs: number;

  constructor(options: WavAudioAnalyzerOptions = {}) {
    this.silenceMarginDb = options.silenceMarginDb ?? 8;
    this.minSilenceMs = options.minSilenceMs ?? 300;
    this.minSpeechMs = options.minSpeechMs ?? 400;
    this.identity = {
      backend: 'wav-statistics',
      model: 'rms-zcr',
      modelVersion: '1',
      locality: 'local',
      mediaLeavesDevice: false,
      parameters: {
        silence_margin_db: this.silenceMarginDb,
        min_silence_ms: this.minSilenceMs,
        min_speech_ms: this.minSpeechMs,
      },
    };
  }

  async analyzeAudio(params: AnalyzeAudioParams): Promise<AnalyzeAudioResult> {
    const stats = computeHopStatistics(params.audio_path, params.hop_ms);
    return analyseHops(stats, {
      silenceThresholdDb: params.silence_threshold_db,
      silenceMarginDb: this.silenceMarginDb,
      minSilenceMs: this.minSilenceMs,
      minSpeechMs: this.minSpeechMs,
    });
  }
}

/**
 * Smallest spread between the quiet and loud parts of a recording for which an
 * adaptive threshold means anything. Below it, the recording has no structure to
 * measure — constant tone, constant hiss, digital silence — and guessing from it
 * produces nonsense.
 */
const MIN_DYNAMIC_RANGE_DB = 10;

/**
 * Where silence begins, for this recording.
 *
 * A fixed -40 dB is wrong in both directions: a quiet indoor recording never
 * reaches it and reads as silent throughout, and a windy street never drops
 * below it and reads as continuous sound. So the threshold is derived from the
 * recording's own floor and spread, and the configured value is kept as the
 * fallback for recordings too flat to measure.
 */
export function hasDynamicRange(rmsDb: number[]): boolean {
  return percentile(rmsDb, 0.9) - percentile(rmsDb, 0.1) >= MIN_DYNAMIC_RANGE_DB;
}

export function silenceThreshold(rmsDb: number[], options: AnalyseHopsOptions): number {
  const floor = percentile(rmsDb, 0.1);
  const ceiling = percentile(rmsDb, 0.9);
  const range = ceiling - floor;
  if (range < MIN_DYNAMIC_RANGE_DB) {
    // Nothing to learn from: fall back, and stay below the floor so a flat
    // recording is not declared silent from end to end.
    return Math.min(options.silenceThresholdDb, floor - 1);
  }
  // Never more than a third of the way up the recording's own range, so a
  // generous margin cannot swallow quiet speech.
  return floor + Math.min(options.silenceMarginDb, range * 0.3);
}

export interface AnalyseHopsOptions {
  silenceThresholdDb: number;
  silenceMarginDb: number;
  minSilenceMs: number;
  minSpeechMs: number;
}

/** Split out from I/O so the whole decision surface is testable with plain arrays. */
export function analyseHops(stats: HopStatistics, options: AnalyseHopsOptions): AnalyzeAudioResult {
  const { rmsDb, zcr, hopMs } = stats;

  const threshold = silenceThreshold(rmsDb, options);
  const measurable = hasDynamicRange(rmsDb);

  // With no dynamic range there is nothing to measure: a constant tone, a
  // constant hiss and a muted track all look identical to an energy detector,
  // and none of them is speech. Reporting zero is the honest answer; a
  // transcriber will say otherwise if there is anything to hear.
  const speechProb = measurable
    ? rmsDb.map((db, i) => speechProbability(db, zcr[i] ?? 0, threshold))
    : rmsDb.map(() => 0);

  const events: AnalyzeAudioResult['events'] = [];
  for (const run of runsOf(rmsDb.map((db) => db < threshold))) {
    const durationMs = (run.end - run.start) * hopMs;
    if (durationMs >= options.minSilenceMs) {
      events.push({
        start_ms: run.start * hopMs,
        end_ms: run.end * hopMs,
        event_type: 'silence',
        confidence: 0.8,
      });
    }
  }
  for (const run of runsOf(speechProb.map((p) => p >= 0.5))) {
    const durationMs = (run.end - run.start) * hopMs;
    if (durationMs >= options.minSpeechMs) {
      events.push({
        start_ms: run.start * hopMs,
        end_ms: run.end * hopMs,
        event_type: 'speech',
        // Energy and zero-crossing rate are a weak speech detector. Saying so in
        // the confidence is better than pretending otherwise: an ASR pass will
        // overwrite this with something far better when one is available.
        confidence: 0.5,
      });
    }
  }
  events.sort((a, b) => a.start_ms - b.start_ms || compareText(a.event_type, b.event_type));

  return {
    model: 'rms-zcr',
    hop_ms: hopMs,
    rms_db: rmsDb.map((v) => round(v, 2)),
    speech_prob: speechProb.map((v) => round(v, 3)),
    events,
  };
}

/**
 * Voiced speech at 16 kHz sits in a narrow zero-crossing band. Below it is hum
 * and rumble, above it is hiss, sibilance and cymbals.
 */
const ZCR_LOW = 0.01;
const ZCR_HIGH = 0.3;

export function speechProbability(rmsDb: number, zcr: number, thresholdDb: number): number {
  if (rmsDb < thresholdDb) return 0;
  const loudness = clamp01((rmsDb - thresholdDb) / 18);
  const band = zcr >= ZCR_LOW && zcr <= ZCR_HIGH ? 1 : zcr < ZCR_LOW ? 0.3 : 0.4;
  return round(clamp01(loudness * band), 3);
}

export function percentile(values: number[], q: number): number {
  if (values.length === 0) return -100;
  const sorted = [...values].sort((a, b) => a - b);
  const index = Math.min(sorted.length - 1, Math.max(0, Math.round((sorted.length - 1) * q)));
  return sorted[index] ?? -100;
}

export function runsOf(flags: boolean[]): { start: number; end: number }[] {
  const runs: { start: number; end: number }[] = [];
  let start: number | null = null;
  for (let i = 0; i < flags.length; i++) {
    if (flags[i]) {
      if (start === null) start = i;
    } else if (start !== null) {
      runs.push({ start, end: i });
      start = null;
    }
  }
  if (start !== null) runs.push({ start, end: flags.length });
  return runs;
}

function clamp01(v: number): number {
  return Math.min(1, Math.max(0, v));
}

function round(value: number, digits: number): number {
  const factor = 10 ** digits;
  return Math.round(value * factor) / factor;
}

File: services/perception/src/editorial_perception/backends/audio.py
"""Loudness, silence and speech presence, from the prepared WAV.

The same algorithm as the TypeScript implementation, deliberately: which one
runs is a deployment detail, and an Editorial IR should not change because the
Python runtime happened to be installed.

It stops short of classifying laughter, applause or music, which need a trained
audio tagger. What it does cover is where the quiet is, which is what lets a cut
land between words instead of through one.
"""

from __future__ import annotations

import math
import wave
from typing import Any

MIN_DYNAMIC_RANGE_DB = 10.0
ZCR_LOW = 0.01
ZCR_HIGH = 0.3


def analyze(
    audio_path: str,
    hop_ms: int = 100,
    silence_threshold_db: float = -40.0,
    *,
    silence_margin_db: float = 8.0,
    min_silence_ms: int = 300,
    min_speech_ms: int = 400,
) -> dict[str, Any]:
    rms_db, zcr = _hop_statistics(audio_path, hop_ms)
    threshold = silence_threshold(rms_db, silence_threshold_db, silence_margin_db)
    measurable = has_dynamic_range(rms_db)

    # With no dynamic range there is nothing to measure: a constant tone, a
    # constant hiss and a muted track look identical to an energy detector, and
    # none of them is speech. Reporting zero is the honest answer.
    speech_prob = (
        [speech_probability(db, z, threshold) for db, z in zip(rms_db, zcr, strict=False)]
        if measurable
        else [0.0] * len(rms_db)
    )

    events: list[dict[str, Any]] = []
    for start, end in _runs([db < threshold for db in rms_db]):
        if (end - start) * hop_ms >= min_silence_ms:
            events.append(
                {
                    "start_ms": start * hop_ms,
                    "end_ms": end * hop_ms,
                    "event_type": "silence",
                    "confidence": 0.8,
                }
            )
    for start, end in _runs([p >= 0.5 for p in speech_prob]):
        if (end - start) * hop_ms >= min_speech_ms:
            events.append(
                # Energy and zero crossings are a weak speech detector, and the
                # confidence says so: a transcriber will overwrite this.
                {
                    "start_ms": start * hop_ms,
                    "end_ms": end * hop_ms,
                    "event_type": "speech",
                    "confidence": 0.5,
                }
            )
    events.sort(key=lambda event: (event["start_ms"], event["event_type"]))

    return {
        "model": "rms-zcr",
        "hop_ms": hop_ms,
        "rms_db": [round(db, 2) for db in rms_db],
        "speech_prob": [round(p, 3) for p in speech_prob],
        "events": events,
    }


def silence_threshold(rms_db: list[float], configured_db: float, margin_db: float) -> float:
    """Where silence begins, for this recording.

    A fixed threshold is wrong in both directions: a quiet indoor recording never
    reaches it and reads as silent throughout, and a windy street never drops
    below it and reads as continuous sound.
    """
    floor = percentile(rms_db, 0.1)
    ceiling = percentile(rms_db, 0.9)
    if ceiling - floor < MIN_DYNAMIC_RANGE_DB:
        return min(configured_db, floor - 1)
    return floor + min(margin_db, (ceiling - floor) * 0.3)


def has_dynamic_range(rms_db: list[float]) -> bool:
    return percentile(rms_db, 0.9) - percentile(rms_db, 0.1) >= MIN_DYNAMIC_RANGE_DB


def speech_probability(rms_db: float, zcr: float, threshold_db: float) -> float:
    if rms_db < threshold_db:
        return 0.0
    loudness = max(0.0, min(1.0, (rms_db - threshold_db) / 18))
    band = 1.0 if ZCR_LOW <= zcr <= ZCR_HIGH else (0.3 if zcr < ZCR_LOW else 0.4)
    return round(max(0.0, min(1.0, loudness * band)), 3)


def percentile(values: list[float], q: float) -> float:
    if not values:
        return -100.0
    ordered = sorted(values)
    index = min(len(ordered) - 1, max(0, round((len(ordered) - 1) * q)))
    return ordered[index]


def _runs(flags: list[bool]) -> list[tuple[int, int]]:
    runs: list[tuple[int, int]] = []
    start: int | None = None
    for index, flag in enumerate(flags):
        if flag and start is None:
            start = index
        elif not flag and start is not None:
            runs.append((start, index))
            start = None
    if start is not None:
        runs.append((start, len(flags)))
    return runs


def _hop_statistics(path: str, hop_ms: int) -> tuple[list[float], list[float]]:
    """Per-hop loudness and zero-crossing rate, in one pass.

    Read in blocks so an hour of audio costs a megabyte of memory rather than a
    hundred.
    """
    with wave.open(path, "rb") as source:
        if source.getsampwidth() != 2:
            raise ValueError(f"expected 16-bit PCM, got {source.getsampwidth() * 8}-bit")
        channels = source.getnchannels()
        rate = source.getframerate()
        frames_per_hop = max(1, round(rate * hop_ms / 1000))

        rms_db: list[float] = []
        zcr: list[float] = []
        sum_squares = 0.0
        crossings = 0
        in_hop = 0
        previous = 0.0
        have_previous = False

        while True:
            block = source.readframes(frames_per_hop * 16)
            if not block:
                break
            for offset in range(0, len(block) - channels * 2 + 1, channels * 2):
                total = 0
                for channel in range(channels):
                    start = offset + channel * 2
                    total += int.from_bytes(block[start : start + 2], "little", signed=True)
                sample = total / channels / 32768

                sum_squares += sample * sample
                if have_previous and (previous >= 0) != (sample >= 0):
                    crossings += 1
                previous = sample
                have_previous = True
                in_hop += 1

                if in_hop == frames_per_hop:
                    rms_db.append(_to_db(math.sqrt(sum_squares / in_hop)))
                    zcr.append(crossings / in_hop)
                    sum_squares = 0.0
                    crossings = 0
                    in_hop = 0

        if in_hop > 0:
            rms_db.append(_to_db(math.sqrt(sum_squares / in_hop)))
            zcr.append(crossings / in_hop)

    return rms_db, zcr


def _to_db(amplitude: float) -> float:
    if amplitude <= 1e-10:
        return -100.0
    return max(-100.0, 20 * math.log10(amplitude))
